mesh/mesh.c

```diff
@@ -470,251 +401,96 @@ static const char *prov_status_str(uint8_t status)
 	}
 }
 
-static void send_join_failed(const char *owner, const char *path,
-							uint8_t status)
-{
-	struct l_dbus_message *msg;
-	struct l_dbus *dbus = dbus_get_bus();
-
-	msg = l_dbus_message_new_method_call(dbus, owner, path,
-						MESH_APPLICATION_INTERFACE,
-						"JoinFailed");
-
-	l_dbus_message_set_arguments(msg, "s", prov_status_str(status));
-	l_dbus_send(dbus_get_bus(), msg);
-
-	free_pending_join_call(true);
-}
-
-static bool prov_complete_cb(void *user_data, uint8_t status,
-					struct mesh_prov_node_info *info)
+static struct l_dbus_message *create_node_call(struct l_dbus *dbus,
+						struct l_dbus_message *msg,
+						void *user_data)
 {
-	struct l_dbus *dbus = dbus_get_bus();
-	struct l_dbus_message *msg;
-	const char *owner;
-	const char *path;
-	const uint8_t *dev_key;
-
-	l_debug("Provisioning complete %s", prov_status_str(status));
-
-	if (!join_pending)
-		return false;
-
-	owner = join_pending->sender;
-	path = join_pending->app_path;
-
-	if (status == PROV_ERR_SUCCESS &&
-	    !node_add_pending_local(join_pending->node, info, mesh.io))
-		status = PROV_ERR_UNEXPECTED_ERR;
-
-	if (status != PROV_ERR_SUCCESS) {
-		send_join_failed(owner, path, status);
-		return false;
-	}
-
-	dev_key = node_get_device_key(join_pending->node);
-
-	msg = l_dbus_message_new_method_call(dbus, owner, path,
-						MESH_APPLICATION_INTERFACE,
-						"JoinComplete");
-
-	l_dbus_message_set_arguments(msg, "t", l_get_u64(dev_key));
-
-	l_dbus_send(dbus_get_bus(), msg);
+	struct request_node_data *new_node;
+	uint16_t *model;
+	struct l_dbus_message_iter iter_uuid, iter_element_models, iter_temp_element;
+	uint32_t temp_element_idx = 0;
+	uint16_t element_idx;
+	bool has_next_model = true;
+	struct l_dbus_message *reply;
+	struct l_queue *element_models;
 
-	free_pending_join_call(false);
+	l_debug("Create node request");
 
-	return true;
-}
+	if (!l_dbus_message_get_arguments(msg, "aya{qaq}", &iter_uuid,
+								&iter_element_models))
+		return dbus_error(msg, MESH_ERROR_INVALID_ARGS, NULL);
 
-static void node_init_cb(struct mesh_node *node, struct mesh_agent *agent)
-{
-	struct l_dbus_message *reply;
-	uint8_t num_ele;
+	new_node = l_new(struct request_node_data, 1);
 
-	if (!node) {
-		reply = dbus_error(join_pending->msg, MESH_ERROR_FAILED,
-				"Failed to create node from application");
-		goto fail;
+	if (dbus_get_byte_array(&iter_uuid, new_node->uuid, UUID_LEN) != UUID_LEN) {
+		l_free(new_node);
+		return dbus_error(msg, MESH_ERROR_INVALID_ARGS, "Bad device UUID");
 	}
 
-	join_pending->node = node;
-	num_ele = node_get_num_elements(node);
+	if (!new_node->elements)
+		new_node->elements = l_queue_new();
 
-	if (!acceptor_start(num_ele, join_pending->uuid, mesh.algorithms,
-				mesh.prov_timeout, agent, prov_complete_cb,
-				&mesh))
+	while (l_dbus_message_iter_next_entry(&iter_element_models, &element_idx,
```

Comment:
We already discussed this loop. Get rid of `has_next_model`, use `l_memdup` for models, etc.